[New Workflow] Flye_denovo to replace DragonFlye

fraser-combe opened the pull request:
This PR closes #611, closes #585, and closes #565.

🗑️ This dev branch should be deleted after merging to main.

## :brain: Summary
This PR introduces a new `flye_denovo` workflow as a replacement for the `Dragonflye` workflow. The updated workflow streamlines the assembly and polishing pipeline, focusing on being flexible and modular with the addition of assembly visualization through Bandage plots. 

Notable enhancements include:

New -tasks, including optional read trimming with `Porechop`, enhanced assembly visualization with `Bandage`, and multiple polishing options. Supports ONT data, hybrid assemblies with Illumina reads, and multiple assembly polishing tools (`Medaka`, `Racon`, and `Polypolish`).
Medaka polishing is set at 1 round as recommended by Rwick, and ONT

## :zap: Impacted Workflows/Tasks
- New `flye_denovo` workflow.
- Replaces and enhances functionality previously offered by the `Dragonflye` workflow.
- Tasks impacted: 
  - `task_porechop.wdl`
  - `task_flye.wdl`
  - `task_bandageplot.wdl`
  - `task_bwa.wdl`
  - `task_medaka.wdl`
  - `task_racon.wdl`
  - `task_dnaapler.wdl`
  - `task_polypolish.wdl`
  - `task_filtercontigs.wdl`
removes task_dragonfly.wdl

This PR may lead to different results in pre-existing outputs: **Yes**

This PR uses an element that could cause duplicate runs to have different results: **Yes**
- Due to the introduction of optional polishing tools and enhancements in assembly parameters, output may vary based on selected configurations.
- This includes updated medaka (including most recent models), polypolish and racon polishing tools from Dragonflye versions
- Updated dnaapler for contig reoirientation - faster run time tested for similar results by authors of the tool

## :hammer_and_wrench: Changes
- Added `flye_denovo.wdl` to replace `Dragonflye`. as a sub workflow
- Enhanced modularity and task-level input definitions for flexibility.
- Integrated multiple polishing and trimming options.
- Introduced better documentation and metadata outputs for transparency and reproducibility.
### :gear: Algorithm

1. **Workflow Redesign:** The `flye_denovo` workflow replaces the `Dragonflye` workflow, with a modular and flexible structure that separates tasks like trimming, assembly, polishing, and final orientation for clarity and maintainability.
2. **Polishing Enhancements:**
   - Added support for **Medaka**, **Racon**, and **Polypolish** with configurable rounds of polishing and tool-specific parameters.
   - Support for hybrid assemblies using Illumina data with `Polypolish`.
3. **Medaka Model Selection:**
   - Introduced automatic Medaka model selection based on the input reads or user-provided overrides.
   - defaults to a medaka model if auto fails otherwise user can override
   - Outputs the Medaka model used
4. **Version Tracking:**
   - Outputs versions of `Flye`, `Porechop`, `Medaka`, `Racon`, `Polypolish`, `Bandage`, and `Dnaapler`.
5. **Outputs:**
   - Outputs now include:
     - Final polished assembly.
     - Bandage plots for graph visualization.
     - Assembly graphs in GFA format.
     - Metadata for task versions
7. **Docker Updates:**
   - Updated Docker images for `Flye`, `Medaka`, `Racon`, `dnaapler` and other tasks to their latest stable versions 

### ➡️ Inputs
No

### ⬅️ Outputs
Added bandage plot png output
version outputs for task level software
medaka models used
Assembly_fasta output from dnaapler for downstream analyses

## :test_tube: Testing
<!-- Please describe how you tested this PR. -->

Scenarios tested within TheiaProk - Expected TheiaProk workflow to complete successfully for each task and specifically for flye_denovo workflow we expect to see successful creation of assembly fasta after any filtering or polishing conducted. 
1. Default path Flye>Medaka Polish>Filtercontige>dnaApler
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/586af547-03dd-4cb8-8877-8041d0064464
medaka output model and version
![image](https://github.com/user-attachments/assets/d57a53f4-8225-4eea-8c6b-63e08aa5efcf)

2. Porechop run i.e skip_trim_reads = false 
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/e5a645db-0e17-4c58-84ce-4e1f44ef9042
3. Skip polishing skip_polishing = true
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/110a4ffa-c208-4849-a3b2-11d88ffddc90

4. Racon polishing pathway (polishing_rounds = 2)
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/95694506-946f-4ef7-9b3d-657522cc7809

5. Hybrid assembly ONT data and Illumina (Polypolish and BWA)
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/4fa9c915-4572-43b9-bd9b-28bed40c75a4

## Comparisons between DragonFlye and New Flye_denovo subworkflow ##
Here we are looking for similarities in assemblies, statistics and downstream analyses. 44 bacterial samples selected

Both workflows produce assemblies of similar lengths for each sample, with minor variations.

Both workflows achieve high BUSCO completeness scores, generally above 90%. 
For those that did not produce above 90% BUSCO scores, the scores were comparable if not better with Flye Denovo or within acceptable ranges.

Both workflows consistently predict the same taxa for each sample.

Previously included BUSCO comparison tables are incorrect and no longer apply to current results, thus were removed. The TheiaValidate run linked below is up to date. 

Downstream analyses
-Gambit Taxon: Identical predictions across workflows for each sample.
 -Both workflows produce identical results for most downstream analyses, ensuring reliable serotype predictions, taxonomic classifications, and virulence gene identifications.

Finally the 44 validation ONT raw data samples were ran through Flye denovo and samples were checked manually to compare against previously ran Dragonflye submissions and we found similar comparable results
https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Hale_Sandbox/submission_history/97fca98f-9f81-41c1-8178-a9301d6ecd72

There were 11 samples that were flagged via their BUSCO scores being discordant by 3% or more, however, the differences were deemed within passable range or displayed an increase in quality from Dragonflye to Flye_denovo. The automated BUSCO check still requires manual interpretation.

### Suggested Scenarios for Reviewer to Test
Parameters to test:
skip_trim_reads: true
skip_polishing: false
polishing_rounds: 1

Expected outputs: Final polished assembly in FASTA format.
Metadata output for versions used (e.g., Flye, Medaka).
No trimming or filtering applied.
Successful Bandage plot and GFA graph generation.

## :microscope: Final Developer Checklist
<!-- Please mark boxes [X] -->
- [x] The workflow/task has been tested and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (Theiagen developers)
- [X] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [X] Documentation and/or workflow diagrams have been updated if applicable
  - [x] You have updated the latest version for any affected worklows in the respective workflow documentation page and for every entry in the three `workflows_overview` tables.

## 🎯 Reviewer Checklist
<!--  Indicate NA when not applicable  -->
- [x] All changed results have been confirmed
- [x] You have tested the PR appropriately (see the [testing guide](https://theiagen.notion.site/PR-Testing-Guide-Determining-Appropriate-Levels-of-Testing-4764e98a6aeb460185039c0896714590) for more information)
- [x] All code adheres to the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-51b66a47dde54c798f35d673fff80249)
- [x] MD5 sums have been updated
- [x] The PR author has addressed all comments
- [x] The documentation has been updated

AndrewLangvt reviewed (changes requested):
Thanks for the work on this @fraser-combe. A few things to address, but we're on our way to unwrapping dragonflye!

sage-wright commented on tasks/assembly/task_flye.wdl:
```diff
@@ -0,0 +1,90 @@
+    set -euo pipefail
+    flye --version | tee VERSION
+
+    # determine read type
```
it really is.

this is my fault, but I'm thinking we can probably smush all the READ_TYPE vars into a single one "read_type" and have it be free text with the default set to "--nano-raw" and then we can list out the different options in the documentation so we don't have like five new variables we just have one

sage-wright commented on tasks/assembly/task_dnaapler.wdl:
```diff
@@ -0,0 +1,72 @@
+
+    # Check input FASTA is valid
+    echo "Validating input FASTA file..."
+    if ! grep -q "^>" ~{input_fasta}; then
```
i think this should be removed entirely, it's unnecessary since this is coming from previous tasks of ours

fraser-combe replied:
I agree makes sense to remove

sage-wright commented on tasks/alignment/task_bwa.wdl:
```diff
@@ -153,4 +153,47 @@ task bwa {
\ No newline at end of file
+}
+
+task bwa_all {
```
that's because there's already a `bwa` task, but this one uses the `all` command which is different enough to warrant its own wdl. i wrote that, and couldn't think of a better name. happy to suggestions here

fraser-combe commented:
Thanks for all the comments

- Updated filter_contigs.wdl - now uses Biopython instead of awk and updated output file text
- Flye- uses read_type input - set default to -nano-hq (open to thoughts to go back to nano-raw) -seems most relavant for recent ONT data going forward
- Ensured runtime parameters standardized across tasks and names
- Updated documentation


Local Testing: Successfully reran the workflow locally with default settings to confirm all tasks executed as expected. Confirmed filter_contigs new workflow output and tested with homopolymer and short read data to confirm trimming

Terra Testing: Reran the workflow in Terra using default settings. No major task overhauls were performed; results matched expectations for typical use cases.

[rerun flye default](https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/31815cdf-d3b3-4e2b-bdb4-34817cef718e)

[reran test after merge ](https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_FCombe_sandbox/job_history/4ab20b28-aea5-47f4-ab64-682f6abee3e8/232d5483-c2d2-408d-94f6-43770b927d8b)

AndrewLangvt commented on tasks/quality_control/read_filtering/task_filter_contigs.wdl:
```diff
@@ -0,0 +1,75 @@
+  input {
+    File assembly_fasta
+    Int min_length = 1000
+    Int disk_size = 10
```
I know @sage-wright asked you to drop the resource allocations here, which I agree with. That said, disk is *relatively cheap compared to CPU/MEM. I'd opt for a bit more here to ensure it doesn't get saturated by the size of some bacterial genomes (or even fungal down the road). Maybe bump this to 50; looks like this might be the general trend across PHB. @sage-wright does that sound accurate?

AndrewLangvt commented on workflows/utilities/wf_flye_denovo.wdl:
```diff
@@ -0,0 +1,222 @@
+    Boolean flye_no_alt_contigs = false
+    Boolean flye_scaffold = false
+    String? flye_additional_parameters # Any extra Flye-specific parameters
+    Int flye_cpu = 4
```
Don't necessarily have to remove these defaults, but as they're specified at the task level, defining them here is redundant. I.e. functionally fine, but you could also just list the input without a default value. Entirely your call, more dropping this comment as food for thought (same for several other inputs here).

AndrewLangvt reviewed (changes requested):
Just a few more to follow up on from Sage's previous comments.

AndrewLangvt commented:
Thanks for making all of these changes @fraser-combe. This looks much more sound, programmatically. I do have a few remaining questions, more on the biological comparison/assessment side of things. I understand the predicted Gambit Taxon was identical across samples assembled with flye vs dragonflye, which is great. It looks like you have a note in this PR to "Add in comparison results." Would you please do that? I think we want to get granular here in how we assess the flye/dragonfle assemblies. In theory, as we're just unwrapping Dragonflye into it's separate components, the assemblies should be largely similar. If you would, please pull together a table with the following (feel free to just update the table that already exists in this PR, if you like:
SNP distance between flye/dragonflye assembled genomes
genome length
BUSCO Scores
predicted taxon
contig length
# contigs
n50

Once you've got this, we can link up as a team to review & make sure our "hive mind" is in agreement across the board.

AndrewLangvt commented:
Thanks for adding the table of SNP differences here @fraser-combe. This shows what I was hoping for - minimal impact to genome assembly. I'm good with merging this PR. I know @sage-wright was taking a look this morn, as well. So, pending any final comments from her, we can get this thing merged!

sage-wright reviewed (commented):
Code looks good! Running a final sanity check [here](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/6c249c57-44a4-48b4-9134-92cd5890acd0) for final confirmation. Will merge upon success!

sage-wright commented on tasks/quality_control/read_filtering/task_filter_contigs.wdl:
```diff
@@ -7,15 +7,13 @@ task filter_contigs {
     Int memory = 8
     Int cpu = 1
-    String docker = "us-docker.pkg.dev/general-theiagen/theiagen/utility:1.2"
+    String docker = "us-docker.pkg.dev/general-theiagen/staphb/kaptive:2.03" # Staph-b docker image with Biopython
```
My tests failed because this docker image doesn't exist. Could you fix this?

fraser-combe replied:
us-docker.pkg.dev/general-theiagen/staphb/kaptive:2.0.3 exists in GAR what tests failed for this docker image?

sage-wright replied:
yeah the second `.` is missing from the tag
"us-docker.pkg.dev/general-theiagen/staphb/kaptive:2.03"
note that it says 2.03 instead of 2.0.3

AndrewLangvt replied:
@fraser-combe can you re-generate the flye assemblies & rerun the comparison between the two? I would imagine this won't significantly impact the resultant assemblies; however, given we're working with the assembly stage of our processes here, we will want to confirm this definitively before merging.

fraser-combe replied:
The run that I used for comparisons the Flye denovo workflow used the old docker image that was successful for filtering contigs. So the results are identical. I reran and confirmed after updating the docker image. None of the samples had any contigs removed because they met the min thresholds of 1000 bp and no homopolymers

sage-wright commented on workflows/theiaprok/wf_theiaprok_ont.wdl:
```diff
@@ -99,16 +99,15 @@ workflow theiaprok_ont {
-           genome_length = select_first([genome_length, read_qc_trim.est_genome_length, 0]),
-           samplename = samplename
+       call flye_workflow.flye_denovo {
+       input:
```
can you fix this indentation while you fix the docker image?

sage-wright commented on workflows/theiaprok/wf_theiaprok_ont.wdl:
```diff
@@ -99,16 +99,15 @@ workflow theiaprok_ont {
-        input:
-          assembly = dragonflye.assembly_fasta,
-          samplename = samplename
+      input:
```
this indentation should be fixed too!

fraser-combe commented:
Updated the docker image name

sage-wright commented:
Thanks for the updates. Running a [TheiaValidate](https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/job_history/5594b8fb-b161-4688-bba8-07a69183c6dd) here for confirmation, will merge upon success 👍

awh082834 commented:
New TheiaValidate run with correct Medaka model being passed to this dev branch. This seems to have cleared up the large discrepancies with the BUSCO scores of previous TheiaValidate runs. I have updated the documentation of the Medaka model for `flye_denovo` to reflect the importance of the model matching the users basecaller. 

Here is the updated [TheiaValidate](https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Hale_Sandbox/submission_history/035e4fc9-64b2-4083-844b-2f1b6e4679cf) run. The number of samples failing the BUSCO comparison has fallen to 21 from 36, however, looking at the exact differences table the results seem fairly close and could be up to algorithmic difference in assembly.

awh082834 commented:
Updated [TheiaValidate](https://app.terra.bio/#workspaces/theiagen-training-workspaces/Theiagen_Hale_Sandbox/submission_history/2c8671d9-3cec-4077-a0c6-166909f4b0d3) run with new output format for the PDF.

Michal-Babins commented on tasks/polishing/task_medaka.wdl:
```diff
@@ -0,0 +1,56 @@
+    medaka --version | tee MEDAKA_VERSION
+
+    # Attempt automatic model resolution if enabled
+    if [[ "~{auto_model}" == "true" ]]; then
```
Thank you Andrew, will resolve once tested.

Michal-Babins commented:
Test with defualts:
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/submission_history/09a72b3c-be28-435e-9db4-bfd32a1346cd

Test with skip polishing set to true:
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/submission_history/ece82f46-75b7-48f3-9a69-39605c5d063b

Test with racon as the polisher:
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/submission_history/5c588e5c-f445-4d9e-b5ee-a6ba0e7c3a6b

Porechop is getting set directly when true:
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/CDPH_Bioinformatics_Development/submission_history/94ac615a-902a-48e3-b83b-238741209de2

Michal-Babins reviewed (approved):
Code is looking good after changes. Theiavalidate is passing and flye is comparable. Tested all differing options. Good to approve now.
